### ISSUE TITLE
[YOLOv4] Discussion

### ISSUE THREAD

**Issue by ElectronicElephant** [open-mmlab/mmdetection#3419]
Hi @hellock and @xvjiarui , 

As the YOLOv3 implementation is going to an end (finally 😄), @WenqiangX and I are planning to work on YOLOv4.

However, compared with YOLOv3, YOLOv4 is notorious for its bags of tricks, many of which are not implemented in `mm-detection` yet. So, we open this issue in the hope of discussing some implementation details with mm-lab.

First, let us summarize the `tricks` used in YOLOv4.

## Already in mmdetection

1. CIoU-Loss
    https://github.com/open-mmlab/mmdetection/pull/3151

2. Label Smoothing
    Included in YOLOv3

3. CosineAnnealingScheduler
    https://github.com/open-mmlab/mmdetection/issues/3227

4. Random training shapes
    Included in Dataloader

## Waiting for merge

1. Mosaic Data Aug
    https://github.com/open-mmlab/mmdetection/pull/3389
    Note that augmentation for mask and segmentation and blur augmentation are not implemented.

2. Efficient-Det
    YOLOv4 **may** rely on the neck of EfficientDet.

## Not implemented yet

- [ ] CutMix

- [ ] DropBlock regularization

- [ ] Mish activation

- [ ] CSP

- [ ] Multi-input weighted residual connections

- [ ] CmBN

- [ ] Self Adversarial Training

- [ ] Eliminate grid sensitivity

- [x] Multiple anchors for a single gt
    Just change the code in YOLOv3Head. Should not be a problem.

- [ ] Optimal Hyper-parameters

- [x] SPP
     Should not be a problem.

- [ ] SAM-block

- [ ] PAN path-aggregation block

- [ ] DIoU-NMS

# Our concern

1. Dependency.

If there is an ETA of EfficientDet and https://github.com/open-mmlab/mmdetection/pull/3389 ?

2. `C++` or `PyTorch` for some tricks, including `DIoU-NMS`.

Should it be adopted in C++ code in `mmcv`, or we can use `PyTorch` code instead (which might be slower)?

3. Customized activation functions.

If we stick to use `ConvModule` of `mmcv`, how to use `Mish-activation` in an elegant manner?

---

Since the implementation of YOLOv4 is still at an early stage, the concern might grow, and we are glad to hear the voice from mm-lab. Cheers.

**Comment by hellock** [comment]
- EfficientDet still have a long way to go since the performance has not been reproduced. Mosaic augmentation should be merged in one (or two) week.
- I prefer implementing DIoU-NMS  with c++ in mmcv, but if the speed difference is not much, python codes are also acceptable.
- It is simple to implement a new activation function in mmcv, like [HSwish](https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/hswish.py#L7). ConvModule supports new activation methods without modifying the source code.

**Comment by hiyyg** [comment]
Hi @ElectronicElephant, I am also interested in having yolov4 in mmdetection. How is it going now?

**Comment by WenqiangX** [comment]
Hi @hiyyg we will pull a request soon (maybe at the end of Sep.), most of the functionality has been implemented.

**Comment by GreenTeaHua** [comment]
great, mark.

**Comment by WenqiangX** [comment]
Due to the computational resources limitation, we are not going to PR the YOLOv4 as planned. But we notice that an official implemention is merged in https://github.com/open-mmlab/mmdetection/pull/3824. @GreenTeaHua
